Add support for psetv2

altafan opened the pull request:
This contribs to the initial effort of @sekulicd in #184 and takes the psetv2 package to a ready-to-use version.

I made a huge refactor regarding named variables, errors messages, and, more importantly, the business logic.

In particular, the blinder role has been basically split into 2 fundamental parts:
* `psetv2.Blinder` is the role in charge of adding blinding data to pset's input issuances or outputs. It makes sure that everything's correct and updates also the global scalars of the partial transaction
* `confidential.zkpGenerator` and `confidential.zkpValidator` are the implementations of `psetv2.BlindingGenerator` and `psetv2.BlindingValidator`, in charge of handling private keys to unblind inputs and public keys to generate blinding data, ie. blinders, commitments, and proofs 

The package already contains some tests showing how to use psetv2 to:
* create a fully unblinded tx
* create a fully unblinded issuance tx
* create an issuance tx with unblinded issuance and blinded outs
* create a fully blinded issuance tx
* create a fully blinded tx
* create a fully blinded swap tx

Please @tiero @sekulicd @louisinger review this.

This closes #167 
This closes #189

altafan commented:
Minor: I added some (tested) util methods to the `elementsutils` package and renamed the existing ones. They all come from projects using this lib. There, we noticed both the usefulness and the lack of such methods, so this seemed a gold opportunity to add them all.
Shouldn't be that problematic fixing their usage in other projects once updating go-elements to the next version.

tiero reviewed (changes requested):
I would prefer to rename `blinder.Blind` into `blinder.AddBlinders` or even split it up into `blinder.AddIssuanceBlinders` and `blinder.AddOutputBlinders` to keep the methods shorters if possible.

Also I am more fan of two separated method `AddOutputBlindersNotLast` and `AddOutputBlindersLast`, what do you think?

tiero commented:
> I would prefer to rename `blinder.Blind` into `blinder.AddBlinders` or eventually split it up into two methods like `blinder.AddIssuanceBlinders` and `blinder.AddOutputBlinders` to keep the methods shorters if possible.
> 
> Also I am more fan of two separated method `AddOutputBlindersNotLast` and `AddOutputBlindersLast` instead of an implicit boolean, what do you think?

Understood splitting Issuance and Output blinders can be problematic with state, but would split `last` and `nonLast` methods?

sekulicd reviewed (commented):
I suggest that in future we do changes over existing PRs, since this way we lose commit and contributor history and its harder to review.
I dont understnad why it is important to extract some blind/sign logic from pset pkg itself. 
Looks like we are trying only to validate(which is not spec req) added blindings and signatures by pset pkg, by the way this is not fully true since we, for eg., do calculation of scalars in blinding.
I find API this way defined, harder to use and it requires more PSET knowledge from pkg user.
Wouldnt be nicer that we ask for required inputs, needed for blinding and signing, and just have one method Blind() and Sing()
in compared to now where client needs more stuff to do eg:
Blinding:
````
        handler := confidential.NewBlinderHandlerFromBlindingKeys(
		[][]byte{blindingPrivateKey.Serialize()},
		nil,
	)

	ownedInputs, err := handler.UnblindInputs(ptx, nil)
	require.NoError(t, err)

	blinder, err := psetv2.NewBlinder(ptx, ownedInputs, handler)
	require.NoError(t, err)

	outBlindingArgs, err := handler.BlindOutputs(ptx, []uint32{1}, nil)
	require.NoError(t, err)

	err = blinder.BlindLast(nil, outBlindingArgs)
	require.NoError(t, err)  
````
Singing
````
	signer, err := psetv2.NewSigner(p)
	if err != nil {
		return err
	}

	for k, v := range p.Inputs {
		if err := signer.AddInSighashType(
			txscript.SigHashAll|transaction.SighashRangeproof, 0,
		); err != nil {
			return err
		}

		prevout := v.GetUtxo()
		prvkey := privKeys[k]
		pubkey := prvkey.PubKey()
		script := scripts[k]

		var sigHash [32]byte
		tx, err := p.UnsignedTx()
		if err != nil {
			return err
		}

		if forWitness {
			sigHash = tx.HashForWitnessV0(
				k,
				script,
				prevout.Value,
				txscript.SigHashAll|transaction.SighashRangeproof,
			)
		} else {
			sigHash, err = tx.HashForSignature(k, script, txscript.SigHashAll|transaction.SighashRangeproof)
			if err != nil {
				return err
			}
		}

		sig, err := prvkey.Sign(sigHash[:])
		if err != nil {
			return err
		}
		sigWithHashType := append(sig.Serialize(), byte(txscript.SigHashAll|transaction.SighashRangeproof))

		var witPubkeyScript, witScript []byte
		if opts != nil {
			witPubkeyScript = opts.pubkeyScript
			witScript = opts.script
		}

		if err := signer.SignInput(
			k,
			sigWithHashType,
			pubkey.SerializeCompressed(),
			witPubkeyScript,
			witScript,
		); err != nil {
			return err
		}
	}

	valid, err := p.ValidateAllSignatures()
	if err != nil {
		return err
	}
	if !valid {
		return fmt.Errorf("invalid signatures")
	}

````
Remember that we are going to replicate above code in all of our projects and it is not part of any pkg but only test.
One last thing, I think we need to move pset_test.go in other pkg if we want to decouple pset from confidential.

@tiero @altafan

altafan commented on psetv2/bitset.go:
```diff
@@ -0,0 +1,87 @@
+	s[index] = 0
+}
+
+func (s BitSet) Test(index int) bool {
```
The naming of BitSet methods respect the naming of the [native c++ structure](https://www.cplusplus.com/reference/bitset/bitset/).

altafan commented on psetv2/blinder.go:
```diff
@@ -0,0 +1,542 @@
+	}
+
+	for _, ownedIn := range b.OwnedInputs {
+		scalar, err = b.blinderHandler.AddToScalarOffset(
```
The code you shared doesn't actually do what you think. Check [here](https://www.cplusplus.com/reference/algorithm/find/) how find() works for iterators: 

> Find value in range
Returns an iterator to the first element in the range [first,last) that compares equal to val. If no such element is found, the function returns last. 

Checking if iterator is equal to end() means checking if the element has been found in the iterator list actually.

altafan commented:
> I suggest that in future we do changes over existing PRs, since this way we lose commit and contributor history and its harder to review.

My branch adds commits to yours (the one for which you opened #184) in order to not lose any of your contributions indeed. And I also think this was the best way to go because otherwise, in order to add commits to your branch, I would have needed write rights on your repo. This is not something you should want to grant to other users over your repos, in general.

> I dont understnad why it is important to extract some blind/sign logic from pset pkg itself.

Extracting the singing logic from the pset is essential to allow every wallet supporting psetv2 to use this library. Otherwise, following your idea, only wallets allowing to dump private keys and use them manually could use our psetv2 pkg. For sure this is not ideal for the usability of the library.

As you showed with your example, I think it's totally fine that one produces the signatures on his own and then uses our signer to add them to the partial transaction, with the role ensuring that everything's correct when doing such operation.

For blinding, we actually shouldn't need such "extraction". We need it though to make psetv2 not dependent from secp256k1-zkp, which is a CGO porting and cause troubles if we want to compile psetv2 to WASM.

So I decided to isolate blinding key management into the handler, which is dependent from secp256k1-zkp and responsible for generating commitments, proof etc and also to unblind blinded data, just like a wallet handle keys to produce signatures to make a comparison.

The other part is the pset's blinder role which ensures that everything's correct when adding commitments, proofs etc to input issuances or outputs. Furthermore, it takes care of keeping the global scalars up-to-date, which is something very pset related.

altafan commented on psetv2/global.go:
```diff
@@ -0,0 +1,395 @@
+	keyPairs = append(keyPairs, fallbackLocktimeKeyPair)
+
+	inputCount := new(bytes.Buffer)
+	if err := wire.WriteVarInt(inputCount, 0, uint64(g.InputCount)); err != nil {
```
@sekulicd can you check if everyhing's fine now? I didn't find the occurrences at L325 and L369.

sekulicd replied:
[this line](https://github.com/altafan/go-elements/blob/3beb5bf353d6855591f9375b2f58253b23d5ba47/psetv2/global.go#L322) and [this one](https://github.com/altafan/go-elements/blob/3beb5bf353d6855591f9375b2f58253b23d5ba47/psetv2/global.go#L366)

sekulicd commented:
> My branch adds commits to yours (the one for which you opened https://github.com/vulpemventures/go-elements/pull/184) in order to not lose any of your contributions indeed.

Hmm in my IDE i was not able to see history of commits, but now when i check GH history i see them, sorry.

> Extracting the singing logic from the pset is essential to allow every wallet supporting psetv2 to use this library. Otherwise, following your idea, only wallets allowing to dump private keys and use them manually could use our psetv2 pkg. For sure this is not ideal for the usability of the library.

Sorry i dont understand this fully, if wallet wants to sign tx it needs to access to priv key? Can u elaborate more on this and maybe if you have example of a wallet which fits your scenario?

> For blinding, we actually shouldn't need such "extraction". We need it though to make psetv2 not dependent from secp256k1-zkp, which is a CGO porting and cause troubles if we want to compile psetv2 to WASM.

secp256k1-zkp is referenced, in pset pkg, through confidential pkg and that is the reason I introduced this [interface](https://github.com/sekulicd/go-elements/blob/24e643124db10fb3f1d2cbd4aeb29acefcc23f83/psetv2/blinder_role.go#L22) for "confidential" methods, so basically moving out blinder logic only to be able to compile WASM seems not a valid reson.

> One last thing, I think we need to move pset_test.go in other pkg if we want to decouple pset from confidential.

Regarding decoupling from secp256k1-zkp, did u checked this?

altafan commented:
> One last thing, I think we need to move pset_test.go in other pkg if we want to decouple pset from confidential.
> Regarding decoupling from secp256k1-zkp, did u checked this?

The package declared in pset_test.go is `psetv2_test`, which is  not `psetv2`. This lets us import psetv2 as external pkg to have the exact "pov" of the user. Importing `psetv2` in other projects won't import `psetv2_test` too, therefore it's alright.

> Sorry i dont understand this fully, if wallet wants to sign tx it needs to access to priv key? Can u elaborate more on this and maybe if you have example of a wallet which fits your scenario?

In the previous comment, after the blind/sign snippet, you said:

> Remember that we are going to replicate above code in all of our projects and it is not part of any pkg but only test.

This is true only for now, just because every wallet we implemented so far are single key P2WPKH wallets, meaning that they sign the inputs in exactly the same way, the one you posted in the snippet.
If we created a sign method like `Sign(privateKeys [][]byte, inIndexes[]uint32)` implementing your snippet:
1. Only single-key P2WPKH wallets could use it
2. Such wallets must have an API to retrieve the list of private keys to provide to the sing method.

And guess what? The majority of single-key wallets outthere are neither native-segwit, nor they let the user directly manipulate private keys such that this hypothetical snippet is possible:
```
privateKeys := walletSvc.GetPrivateKeys()
signer := psetv2.NewSigner(ptx)
signedPtx :=  signer.Sign(privateKeys, inIndexes)
```

> secp256k1-zkp is referenced, in pset pkg, through confidential pkg and that is the reason I introduced this [interface](https://github.com/sekulicd/go-elements/blob/24e643124db10fb3f1d2cbd4aeb29acefcc23f83/psetv2/blinder_role.go#L22) for "confidential" methods, so basically moving out blinder logic only to be able to compile WASM (@tiero) seems not a valid reason.

I think your initial idea is good indeed, I just tried to improve it in order for the user to have a clear separation of things that require blinding keys from those that do not.

tiero reviewed (changes requested):
the doc.go in pset is needed to self generate it on 

https://pkg.go.dev/github.com/vulpemventures/go-elements@v0.3.7/pset

Would not remove (rather keep it outdated and update later on eventually)

tiero commented on psetv2/utils_test.go:
```diff
@@ -0,0 +1,289 @@
+	p *psetv2.Pset,
+	privKeys []*btcec.PrivateKey,
+	scripts [][]byte,
+	forWitness bool,
```
Woudl specify already version here? To make it self evident for Upcoming taproot update

altafan commented on confidential/blinder_handler.go:
```diff
@@ -0,0 +1,569 @@
+	Rng         RandomNumberGenerator
+	ExtraInputs []UnblindOutputResult
+}
+
```
@tiero done

tiero commented on confidential/confidential.go:
```diff
@@ -47,7 +48,16 @@ func UnblindOutputWithKey(
+		}
+		return &UnblindOutputResult{
+			Value:               value,
+			Asset:               out.Asset[1:],
```
aren't we reversing it?

tiero reviewed (commented):
Update the badge in README for the CI status, to point to Github Action badge

[![Go](https://github.com/vulpemventures/go-elements/actions/workflows/ci.yml/badge.svg)](https://github.com/vulpemventures/go-elements/actions/workflows/ci.yml)

tiero reviewed (changes requested):
Let's add the BIP371 Taproot fields

tiero commented:
https://github.com/btcsuite/btcd/pull/1847

altafan commented on psetv2/creator.go:
```diff
@@ -94,7 +95,10 @@ func (a OutputArgs) validate() error {
 	var script, blindingKey []byte
 	if len(a.Address) > 0 {
-		script, _ = address.ToOutputScript(a.Address)
+		script = []byte{txscript.OP_RETURN}
```
Yes indeed.
`OutputArgs` is a user-friendly type  without a raw field like `BlindingPubKey []byte`. It instead has a `Address string`  field that can be confidential or unconfidential, and in the first case is used to set also the partial output blinding pubkey.
If the OutputArgs has a confidential address and a 0 amount, it means it's dummy confidential output, therefore only the blinding pubkey extracted from the confidential address is passed to the partial output, while the actual script is OP_RETURN.

tiero replied:
bump @altafan
